src/lib.rs

```diff
@@ -0,0 +1,716 @@
+
+    /// The type of FAT filesystem.
+    fat_type: FatFsType,
+
+    /// The count of cluster availaible in the filesystem.
+    cluster_count: u32,
+}
+
+#[allow(dead_code)]
+impl FatVolumeBootRecord {
+    /// Offset of the boot signature.
+    const BOOTABLE_SIGNATURE: usize = 510;
+
+    /// Offset of the FAT system identifier.
+    const SYSTEM_IDENTIFIER_FAT: usize = 54;
+
+    /// Offset of the FAT32 system identifier.
+    const SYSTEM_IDENTIFIER_FAT32: usize = 82;
+
+    /// Create a new FAT volume boot record from raw data.
+    pub fn new(data: [u8; MINIMAL_BLOCK_SIZE]) -> Option<Self> {
+        let mut res = Self::new_unchecked(data);
+
+        if !res.is_valid() {
+            return None;
+        }
+
+        res.initialize_cache();
+
+        Some(res)
+    }
+
+    /// Initialize extra infos computed from the data backend.
+    pub fn initialize_cache(&mut self) {
+        let root_dir_blocks = ((u32::from(self.root_dir_childs_count()) * 32)
+            + (u32::from(self.bytes_per_block()) - 1))
+            / u32::from(self.bytes_per_block());
+        let data_blocks = self.total_blocks()
+            - (u32::from(self.reserved_block_count())
+                + (u32::from(self.fats_count()) * self.fat_size())
+                + root_dir_blocks);
+        let cluster_count = data_blocks / u32::from(self.blocks_per_cluster());
+        if cluster_count < 0xFF5 {
+            self.fat_type = FatFsType::Fat12;
+        } else if cluster_count < 0xFFF5 {
+            self.fat_type = FatFsType::Fat16;
+        } else {
+            self.fat_type = FatFsType::Fat32;
+        }
+        self.cluster_count = cluster_count + 2;
+    }
+
+    /// Create a new FAT volume boot from raw data without checking for validity.
+    pub(crate) fn new_unchecked(data: [u8; MINIMAL_BLOCK_SIZE]) -> Self {
+        FatVolumeBootRecord {
+            data,
+            fat_type: FatFsType::Fat12,
+            cluster_count: 0,
+        }
+    }
+
+    /// Checks the validity of the boot record.
+    pub fn is_valid(&self) -> bool {
+        // check boot signature
+        if LittleEndian::read_u16(
+            &self.data[Self::BOOTABLE_SIGNATURE..Self::BOOTABLE_SIGNATURE + 2],
+        ) != 0xAA55
+        {
+            return false;
+        }
+
+        // check jump code
+        if self.data[0] != 0xE9 && self.data[0] != 0xEB && self.data[0] != 0xE8 {
+            return false;
+        }
+
+        // check system identifier
+        if self.data[Self::SYSTEM_IDENTIFIER_FAT..Self::SYSTEM_IDENTIFIER_FAT + 3] != *b"FAT"
+            && self.data[Self::SYSTEM_IDENTIFIER_FAT32..Self::SYSTEM_IDENTIFIER_FAT32 + 5]
+                != *b"FAT32"
+        {
+            return false;
+        }
+
+        if self.bytes_per_block() < MINIMAL_BLOCK_SIZE as u16 {
+            return false;
+        }
+
+        true
+    }
+
+    /// Mark the boot record as valid
+    pub(crate) fn set_valid(&mut self, fat_type: FatFsType) {
```

Comment:
This is confusing, you have both `fat_type` passed as an argument, and `self.fat_type`. What happens when they're not the same ?

The only place where you use this function is in `format_partition`. The fact that it takes `&mut self` is weird. I believe this should be a constructor, change it to
```suggestion
    pub(crate) fn create(fat_type: FatFsType) -> FatVolumeBootRecord {
```
allocate an empty block, prefill the desired fields, return it, and let `format_partition` fill the remaining fields. 

This removes the need for a `new_unchecked` constructor.

Reply:
The self.fat_type refer to a cached value that isn't initialized during format.

This function only write signatures based on the given fat_type. The fat_type is determined during the formatting so I cannot do a create function containing the entire initialization of the FatVolumeBootRecord.